File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/SkillFrontmatterRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Rule that requires skill files to have proper frontmatter.
 */
public enum SkillFrontmatterRule implements HarnessCheckRule {
    INSTANCE;

    @Override
    public String category() {
        return "skillFrontmatter";
    }
    private static final String CATEGORY = "skillFrontmatter";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final Path directory = root.resolve(".claude/skills");
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        if (!HarnessCheckHelper.isSafeDirectory(root, directory)) {
            return List.of(new Finding(severity, CATEGORY, ".claude/skills must contain at least one SKILL.md"));
        }
        final List<Path> skillFiles = HarnessCheckHelper.safeFileOrWalk(root, directory).stream()
                .filter(f -> f.getFileName().toString().equals("SKILL.md"))
                .toList();
        if (skillFiles.isEmpty()) {
            return List.of(new Finding(severity, CATEGORY, ".claude/skills must contain at least one SKILL.md"));
        }
        final Pattern descPattern = Pattern.compile("(?m)^description:\\s*.+$");
        return skillFiles.stream()
                .flatMap(f -> {
                    try {
                        return validateSkillFile(root, f, descPattern, severity).stream();
                    } catch (MojoExecutionException e) {
                        return Stream.empty();
                    }
                })
                .toList();
    }

    private List<Finding> validateSkillFile(Path root, Path file, Pattern descPattern, String severity) throws MojoExecutionException {
        final String text = HarnessCheckHelper.readFile(root, file);
        final String relative = root.relativize(file).toString();
        return Stream.<Stream<Finding>>of(
                text.startsWith("---") ? Stream.<Finding>empty() : Stream.of(new Finding(severity, CATEGORY, "skill missing frontmatter: " + relative)),
                descPattern.matcher(text).find() ? Stream.<Finding>empty() : Stream.of(new Finding(severity, CATEGORY, "skill missing description: " + relative))
        ).flatMap(s -> s).toList();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/SkillFrontmatterRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.HarnessPsiResults.Finding
import ai.harness.gradle.HarnessCheck
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path
import kotlin.io.path.div
import kotlin.io.path.isDirectory
import kotlin.io.path.isRegularFile
import kotlin.io.path.name
import kotlin.io.path.readText
import kotlin.io.path.relativeTo
import kotlin.io.path.walk

/**
 * Rule that requires skills to have proper frontmatter.
 */
object SkillFrontmatterRule : HarnessCheckRule() {
    override val category: String = "skillFrontmatter"
    override fun applies(manifest: JsonObject): Boolean =
        manifest[category]?.jsonObject?.get("enabled")?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true

    override fun validate(
        manifest: JsonObject,
        root: Path,
    ): Collection<Finding> = buildList {
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        val messagesObj = catObj?.get("messages")?.jsonObject
        if (catObj != null && parametersObj != null && messagesObj != null) {
            val rootDirectory = HarnessCheck.stringFrom(parametersObj, "rootDirectory")
            val filename = HarnessCheck.stringFrom(parametersObj, "filename")
            val dirPath = root / rootDirectory
            val severity = HarnessCheck.severityOf(manifest, category)
            if (!dirPath.isDirectory()) {
                add(
                    Finding(
                        severity,
                        category,
                        HarnessCheck.stringFrom(messagesObj, "missingDirectory").takeIf { message ->
                            message.isNotEmpty()
                        } ?: ".claude/skills must contain at least one SKILL.md",
                    ),
                )
            } else {
                val files =
                    dirPath
                        .walk()
                        .filter { file -> file.isRegularFile() }
                        .filter { file -> file.name == filename }
                        .toList()
                if (files.isEmpty()) {
                    add(
                        Finding(
                            severity,
                            category,
                            HarnessCheck.stringFrom(messagesObj, "missingSkill").takeIf { message ->
                                message.isNotEmpty()
                            } ?: ".claude/skills must contain at least one SKILL.md",
                        ),
                    )
                }
                files.forEach { file ->
                    val text = file.readText()
                    if (!text.startsWith("---")) {
                        add(
                            Finding(
                                severity,
                                category,
                                HarnessCheck
                                    .stringFrom(
                                        messagesObj,
                                        "missingFrontmatter",
                                    ).takeIf { message -> message.isNotEmpty() }
                                    ?: "skill missing frontmatter: ${file.relativeTo(root)}",
                            ),
                        )
                    } else if (!"""(?m)^description:\s*.+$""".toRegex().containsMatchIn(text)) {
                        add(
                            Finding(
                                severity,
                                category,
                                HarnessCheck
                                    .stringFrom(messagesObj, "missingDescription")
                                    .takeIf { message -> message.isNotEmpty() }
                                    ?: "skill missing description: ${file.relativeTo(root)}",
                            ),
                        )
                    }
                }
            }
        }
    }

    override fun renderPsiFindings(
        findings: List<ai.harness.gradle.PsiFinding>,
        manifest: JsonObject,
    ): Collection<Finding> = emptyList()

    override fun findPsiFindings(
        file: Path,
        root: Path,
        psiFactory: org.jetbrains.kotlin.psi.KtPsiFactory?,
    ): List<ai.harness.gradle.PsiFinding> = emptyList()
}
